File: Counting/Filled_Subgrid_Count_I.py
n,k = map(int,input().split())
grid = []
for _ in range(n):
    s = list(map(lambda x : ord(x) - ord('A'),list(input())))
    grid.append(s)
ans = [0]*k
dp =[[0]*n for _ in range(n)]
for i in range(n-1,-1,-1):
    for j in range(n-1,-1,-1):
        a = dp[i + 1][j] if i + 1 < n and grid[i + 1][j]==grid[i][j] else 0
        b = dp[i][j + 1] if j + 1 < n  and grid[i][j + 1]==grid[i][j] else 0
        c = dp[i + 1][j + 1] if i + 1 < n  and j + 1<n and grid[i + 1][j + 1]==grid[i][j]   else 0
        dp[i][j] = 1 + min(a,b,c)
        ans[grid[i][j]]+=dp[i][j]
for i in range(k):print(ans[i])

File: Counting/Filled_Subgrid_Count_I.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

int main() {
    int n, k;
    cin >> n >> k;
    vector<vector<int>> grid(n, vector<int>(n));
    for (int i = 0; i < n; ++i) {
        string s;
        cin >> s;
        for (int j = 0; j < n; ++j) {
            grid[i][j] = s[j] - 'A';
        }
    }
    for (int curr = 0; curr < k; ++curr) {
        long long ans = 0;
        vector<vector<int>> dp(n, vector<int>(n, 0));
        for (int i = n - 1; i >= 0; --i) {
            for (int j = n - 1; j >= 0; --j) {
                if (grid[i][j] == curr) {
                    int a = (i + 1 < n) ? dp[i + 1][j] : 0;
                    int b = (j + 1 < n) ? dp[i][j + 1] : 0;
                    int c = (i + 1 < n && j + 1 < n) ? dp[i + 1][j + 1] : 0;
                    dp[i][j] = 1 + min({a, b, c});
                    ans += dp[i][j];
                }
            }
        }
        cout << ans << endl;
    }
    return 0;
}